pkg/auth/authenticator/password/htpasswd/htpasswd.go

```diff
@@ -0,0 +1,195 @@
+		file:   file,
+		mapper: mapper,
+	}
+	if err := auth.loadIfNeeded(); err != nil {
+		return nil, err
+	}
+	return auth, nil
+}
+
+func (a *Authenticator) AuthenticatePassword(username, password string) (user.Info, bool, error) {
+	a.loadIfNeeded()
+
+	if len(username) > 255 {
+		username = username[:255]
+	}
+	if strings.Contains(username, ":") {
+		return nil, false, errors.New("Usernames may not contain : characters")
+	}
+	hash, ok := a.usernames[username]
+	if !ok {
+		return nil, false, nil
+	}
+	if ok, err := testPassword(password, hash); !ok || err != nil {
+		return nil, false, err
+	}
+
+	identity := &authapi.DefaultUserIdentityInfo{
+		UserName: username,
+	}
+	user, err := a.mapper.UserFor(identity)
+	glog.V(4).Infof("Got userIdentityMapping: %#v", user)
+	if err != nil {
+		return nil, false, fmt.Errorf("Error creating or updating mapping for: %#v due to %v", identity, err)
+	}
+
+	return user, true, nil
+
+}
+
+func (a *Authenticator) load() error {
+	file, err := os.Open(a.file)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	newusernames := map[string]string{}
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if len(line) == 0 {
+			continue
+		}
+
+		parts := strings.SplitN(line, ":", 2)
+		if len(parts) != 2 {
+			glog.Warningf("Ignoring malformed htpasswd line: %s", line)
+			continue
+		}
+
+		username := parts[0]
+		password := parts[1]
+		if _, ok := newusernames[username]; ok {
+			glog.Warningf("htpasswd contains multiple passwords for user: %s", username)
+		}
+
+		newusernames[username] = password
+	}
+	// TODO: need to lock here?
+	a.usernames = newusernames
+
+	return nil
+}
+
+func (a *Authenticator) loadIfNeeded() error {
+	info, err := os.Stat(a.file)
+	if err != nil {
+		return err
+	}
+	if a.fileInfo == nil || a.fileInfo.ModTime() != info.ModTime() {
+		glog.V(4).Infof("Loading htpasswd file %s...", a.file)
+		a.fileInfo = info
+		return a.load()
+	}
+	return nil
+}
+
+func testPassword(password, hash string) (bool, error) {
+	switch {
+	case strings.HasPrefix(hash, "$apr1$"):
+		// MD5, default
+		return testMD5Password(password, hash)
+	case strings.HasPrefix(hash, "$2y$") || strings.HasPrefix(hash, "$2a$"):
+		// Bcrypt, secure
+		return testBCryptPassword(password, hash)
+	case strings.HasPrefix(hash, "{SHA}"):
+		// SHA-1, insecure
+		return testSHAPassword(password, hash[5:])
+	case len(hash) == 13:
+		// looks like crypt
+		return testCryptPassword(password, hash)
+	default:
+		return false, errors.New("Unrecognized hash type")
+	}
+}
+
+func testSHAPassword(password, hash string) (bool, error) {
+	if len(hash) == 0 {
+		return false, errors.New("Invalid SHA hash")
+	}
+	// Compute hash of password
+	shasum := sha1.Sum([]byte(password))
+	// Base-64 encode
+	base64shasum := base64.StdEncoding.EncodeToString(shasum[:])
+	// Compare
+	match := hash == base64shasum
+	return match, nil
+}
+
+func testBCryptPassword(password, hash string) (bool, error) {
```

Comment:
Updated to return errors and re-enabled test cases for those types to make sure all cases result in errors